libraries/platforms/dromajo-vcs/library.mk

```diff
@@ -0,0 +1,166 @@
+LIB_SSOURCES += $(BLACKPARROT_SDK_DIR)/perch/muldiv.S
+LIB_SSOURCES += $(BSG_PLATFORM_PATH)/software/src/crt0.S
+
+# For now, use the noimpl version of the features
+include $(LIBRARIES_PATH)/features/dma/noimpl/feature.mk
+
+$(DMA_FEATURE_OBJECTS): INCLUDES += -I$(BSG_PLATFORM_PATH)/software/include
+$(DMA_FEATURE_OBJECTS): CFLAGS += -march=rv64imafd -mcmodel=medany -mabi=lp64 -D_BSD_SOURCE -D_XOPEN_SOURCE=500
+$(DMA_FEATURE_OBJECTS): CXXFLAGS += -march=rv64imafd -mcmodel=medany -mabi=lp64 -D_BSD_SOURCE -D_XOPEN_SOURCE=500
+
+# Add the riscv-newlib specific includes for the library
+LIB_PLATFORM_INCLUDES  = -I$(BSG_PLATFORM_PATH)
+LIB_PLATFORM_INCLUDES += -I$(BSG_PLATFORM_PATH)/software/include
+LIB_PLATFORM_INCLUDES += -I$(BLACKPARROT_SDK_DIR)/perch
+
+$(LIB_OBJECTS) $(LIB_OBJECTS_CUDA_POD_REPL) $(LIB_OBJECTS_REGRESSION): CC = $(RV_CC)
+$(LIB_OBJECTS) $(LIB_OBJECTS_CUDA_POD_REPL) $(LIB_OBJECTS_REGRESSION): CXX = $(RV_CXX)
+
+$(BSG_PLATFORM_PATH)/libbsg_manycore_runtime.a: AR = $(RV_AR)
+$(BSG_PLATFORM_PATH)/libbsgmc_cuda_legacy_pod_repl.a: AR = $(RV_AR)
+$(BSG_PLATFORM_PATH)/libbsg_manycore_regression.a: AR = $(RV_AR)
+
+# Make the litteFS file system
+$(BSG_PLATFORM_PATH)/lfs.cpp: MKLFS = $(BLACKPARROT_SDK_DIR)/install/bin/dramfs_mklfs 128 64
+$(BSG_PLATFORM_PATH)/lfs.cpp:
+	$(MKLFS) > $@
+
+$(BSG_PLATFORM_PATH)/lfs.o: $(BSG_PLATFORM_PATH)/lfs.cpp
+	$(CXX) $^ -c -o $@ $(CXXFLAGS) $(INCLUDES)
+
+include $(LIBRARIES_PATH)/features/dma/simulation/dramsim3.mk
+include $(LIBRARIES_PATH)/features/dma/simulation/libdmamem.mk
+
+$(LIBRARIES_PATH)/features/dma/simulation/libdmamem.so: INCLUDES :=
+$(LIBRARIES_PATH)/features/dma/simulation/libdramsim3.so: INCLUDES :=
+
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/cutils.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/dromajo_cosim.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/dromajo_main.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/dromajo_manycore.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/dw_apb_uart.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/elf64.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/fs.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/fs_disk.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/iomem.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/json.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/machine.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/pci.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/riscv_cpu.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/riscv_machine.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/softfp.cpp
+DROMAJO_CXXSOURCES += $(DROMAJO_DIR)/src/virtio.cpp
+
+DROMAJO_OBJECTS += $(patsubst %cpp,%o,$(DROMAJO_CXXSOURCES))
+
+$(DROMAJO_OBJECTS): INCLUDES := -I$(DROMAJO_DIR)/include
+$(DROMAJO_OBJECTS): INCLUDES += -I$(BSG_PLATFORM_PATH)/software/include/dromajo
+$(DROMAJO_OBJECTS): CXXFLAGS := -std=c++11 -fPIC -Wall -Wno-parentheses -MMD -D_FILE_OFFSET_BITS=64 -D_LARGEFILE_SOURCE -D_GNU_SOURCE  -D__STDC_FORMAT_MACROS -DFIFO_MAX_ELEMENTS=$(BSG_MACHINE_IO_EP_CREDITS) -std=gnu++11
```

Comment:
IIRC, BSG_MACHINE_IO_EP_CREDITS is defined in Makefile.machine.include. I think you should include that file in this fragment, so that it's clear(er) where this macro comes from.

Unless including it causes a bunch of warnings...

Reply:
Added a include statement here and in compilation.mk. The one here makes it clear that the BSG_MACHINE_IO_EP_CREDITS and BSG_MACHINE_HOST_COORD_* used by the simulator are coming from the machine Makefile. The one in compilation.mk does the same for the RISC-V code.

Reply:
You should make DROMAJO_OBJECTS depend on Makefie.machine.include, too. That way, in the rare case that either variable changes, the relevant objects get rebuilt.